src/main/java/com/google/step/datamanager/DealManagerDatastore.java

```diff
@@ -134,40 +157,136 @@ public Deal updateDeal(Deal deal, List<String> tagNames) {
     }
     datastore.put(dealEntity);
     searchManager.putDeal(deal, dealTagManager.getTagIdsOfDeal(deal.id));
-    return transformEntitytoDeal(dealEntity);
+    return transformEntityToDeal(dealEntity);
+  }
+
+  /** Retrieves deals posted by _ followed by user */
+  private List<Deal> getDealsPublishedByFollowedRestaurantsOrUsers(
+      List<Long> idsOfFollowedFieldName, String filterAttribute) {
+    List<Deal> dealResults = new ArrayList<>();
+    for (Long id : idsOfFollowedFieldName) {
+      Filter propertyFilter = new FilterPredicate(filterAttribute, FilterOperator.EQUAL, id);
+      Query query = new Query("Deal").setFilter(propertyFilter);
+      PreparedQuery pq = datastore.prepare(query);
+      for (Entity entity : pq.asIterable()) {
+        dealResults.add(readDeal(entity.getKey().getId()));
+      }
+    }
+    return dealResults;
   }
 
   /** Retrieves deals posted by users followed by user */
   @Override
   public List<Deal> getDealsPublishedByFollowedUsers(List<Long> userIds) {
-    return new ArrayList<Deal>();
+    return getDealsPublishedByFollowedRestaurantsOrUsers(userIds, "posterId");
   }
 
   /** Retrieves deals posted by restaurants followed by user */
   @Override
   public List<Deal> getDealsPublishedByFollowedRestaurants(List<Long> restaurantIds) {
-    return new ArrayList<Deal>();
+    return getDealsPublishedByFollowedRestaurantsOrUsers(restaurantIds, "restaurantId");
   }
 
   /** Retrieves deals posted by tags followed by user */
   @Override
   public List<Deal> getDealsPublishedByFollowedTags(List<Long> tagIds) {
-    return new ArrayList<Deal>();
+    List<Long> dealIdResults = new ArrayList<>();
+    for (Long id : tagIds) {
+      List<Long> dealIdsWithTag = dealTagManager.getDealIdsWithTag(id);
+      dealIdResults.addAll(dealIdsWithTag);
+    }
+    // Get rid of duplicate dealID (Deals with multiple tags)
+    List<Long> dealsWithoutDuplicates = new ArrayList<>(new HashSet<>(dealIdResults));
+    return readDeals(dealsWithoutDuplicates);
   }
 
+  /** Sorts deals based on votes (Highest to lowest) */
   @Override
   public List<Deal> sortDealsBasedOnVotes(List<Deal> deals) {
```

Comment:
do you want to move this method to the servlet? Because the servlets will need to call getVotes anyway to print the deal details. It seems like there is a lot of extra calls to getVotes.